### ISSUE TITLE
shouldn't embeding be normalized along embed dimension?

### ISSUE THREAD

**Issue by jfpuget** [lucidrains/nGPT-pytorch#9]
In https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L350 token_embed is normalized along last dimension when it should be normalized along the first dimension.

**Comment by lucidrains** [comment]
hey Jean-Francois, it is actually defaulted to first dimension [here](https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L124)

**Comment by jfpuget** [comment]
You are right. My bad, I saw that but somehow mixed with the embedding layout where weight is the transpose of that of linear layer.

**Comment by jfpuget** [comment]
Sorry to bother you again, but aren't to_q, to_k and to_v normalized on the input dim when they should be normalized on the output dim?

https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L196

**Comment by lucidrains** [comment]
just read your deleted message, you must have figured it out. thanks for the code review regardless

**Comment by jfpuget** [comment]
Thanks. I am adapting your code to a different use case than sequence to sequence. I will report if I get anything good out of my experiments.

**Comment by lucidrains** [comment]
@jfpuget sounds good, so far i haven't seen anyone get breakout results, but it is interesting it converges competitively with this amount of normalization. def let me know if you find anything that does not line up with the paper

**Comment by jfpuget** [comment]
first update: the code does not work with mixed precision training (cuda.amp). I'll look into it, but sharing in case you have it working with it.

It runs, but it does not train.

Maybe something to do with eps. I use 0.

**Comment by lucidrains** [comment]
oh interesting, I'll check it out tomorrow morning

you can also try editing the 1e-5 here to 1e-3 https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L47 (just a shot in the dark)

**Comment by jfpuget** [comment]
Tx. I'll look into it tomorrow (past midnight here now).

Training in full precision seems to be working great.

**Comment by lucidrains** [comment]
@jfpuget yea it does converge, but i haven't seen any big gains over non-normalized transformer fwiw

**Comment by lucidrains** [comment]
@jfpuget hey, think it was an issue with the epsilon i was manually managing

could you try again on `0.1.12`?

**Comment by jfpuget** [comment]
Issue is still there. I see you don't use a gradient scaler in your training code. If I remove it from my training code than loss is nan.

The issue maybe in my model. I adapted yours to a 2D case, and maybe I goofed. I'll report if I find why fp16 does not work for me.

**Comment by lucidrains** [comment]
@jfpuget ah I see, thanks!

well it resolves an issue I had on my end but will play around with gradscaler later

**Comment by lucidrains** [comment]
@jfpuget if somehow you are working with some public (vision?) dataset, if you can get me a reproducible script, that would allow me to help you debug this faster

**Comment by lucidrains** [comment]
yea, seems to converge fine even once i have the gradscaler in there. i'll leave the script with mixed precision defaulted on

**Comment by jfpuget** [comment]
I can't share unfortunately, but thanks for the offer!

Another feedback is that the parametrized version works better (for me) than the manual norm.

I reckon it is because when you manually norm, the Adam momentum is not updated. It should be scaled the same way the weight is scaled when normalized.

The parametrized version does it automatically as gradients are computed through it.

I see you set the manual norm as default I your training code. Maybe that's why You don't see the gains other see with the parametrized versions?


Sorry if I sound negative. Your repo is awesome (same for ROPE), and it probably saved me days of tedious debugging! Thanks for putting it together.

**Comment by lucidrains** [comment]
@jfpuget thank you for sharing that, and indeed i have been testing with manual updates only. will do some full reruns when i find a stretch of time

and no problem in re: feedback! this actually helps a lot for my figuring out whether to put more time into an implementation

**Comment by lucidrains** [comment]
@jfpuget got it working?

**Comment by jfpuget** [comment]
mixed precision? No, not trying for now.

I wanted to update on something else but was waiting for my training run to end.

I wonder why you have separate q_scale and k_scale when they are the same in the nGPT paper. I am training with a single scale for both q and k for now.

I also moved the rope before normalization. It should not change much given rotation does not change the norm. But I did it as I think this is what the paper says.

**Comment by lucidrains** [comment]
@jfpuget ah i see, well a phd student in another thread got it working but with bfloat16

i'm following a traditional qk rmsnorm formulation being used by Robin Rombach and his team, who i really trust. but if you notice a big improvement using just one scale, i can change it

**Comment by lucidrains** [comment]
@jfpuget i decided to just make the change so people can build on this research. let me know if you see anything that doesn't look right

**Comment by jfpuget** [comment]
Training still going on. train loss is similar, but valid loss is better with a single scale. I'll report tomorrow on the full train run result.

Your code looks correct to me (moving rope before norm, and single scale). That was fast!

**Comment by lucidrains** [comment]
will def update my beliefs based on your experiments

can tell the architecture is in skilled hands

**Comment by jfpuget** [comment]
Thanks a lot for your trust but I am not sure it is deserved. The upside I saw early in the training seems to vanish. Maybe your way was better actually.

I'll run more experiments but it would be good to have other people feedback on this change maybe.

**Comment by lucidrains** [comment]
@jfpuget ah got it, thanks for the update!

**Comment by jfpuget** [comment]
Another run confirms the change is not very important. training loss and valid loss are slightly better, but could be just due to changes in layer init given the reordering. I'll keep it as it is closer to the paper IMHO, but that's the only reason.

I'll look again at mixed precision in the coming days.

**Comment by jfpuget** [comment]
I found why mixed precision wasn't working. My use case requires a mask in attention., I relied on the Pytorch default for a Boolean mask. But it seems it is not working that well depending on the kernel used for sdpa. I replaced the Boolean mask with a numerical value (-10 for now) for masked positions. Mixed precision now works fine.

I found it because on A100 the sdpa with Boolean mask was leading to Nan gradients. So far I was using a A6000 Ada.

**Comment by lucidrains** [comment]
@jfpuget thanks again! i threw in a quick fix, but that should probably be submitted as an issue upstream

**Comment by jfpuget** [comment]
You mean an issue with pytorch sdpa?

**Comment by jfpuget** [comment]
I will extract the code that shows the issue and open a pytorch issue.

**Comment by lucidrains** [comment]
@jfpuget could it be that you had an entire row that was masked out? that would `NaN` for softmax and would be the correct behavior

**Comment by lucidrains** [comment]
@jfpuget is this on top of autoregressive?

**Comment by jfpuget** [comment]
update. As said in other thread, the fp1q6 is solved. In my use case I do axial attnetion in a 2D setting. Not really a computer vision use case. I'll say more in few weeks.

Another change I tried that seems useful: I normalize logits before scaling them, to stay in the spirit of nGPT. It helps a bit on my use case.

**Comment by lucidrains** [comment]
@jfpuget excited to hear what you trained, once you can say!

**Comment by jfpuget** [comment]
Hi Phil, I can now share what I did. It was for a Kaggle competition. I wrote a short paper, you can see the content there: https://www.kaggle.com/competitions/arc-prize-2024/discussion/545844 . I'll share my code later today.

**Comment by JeanKaddour** [comment]
Just read this entire thread, very interesting, thank you both for sharing your insights. 

@jfpuget thank you for sharing your notebook. It's funny that you used nGPT for ARC because I've been independently dabbling with both but didn't draw the connection.

I was looking into nGPT as part of an [ongoing paper](https://arxiv.org/abs/2307.06440) of mine. I couldn't materialize any gains compared to the Decoder-only Transformer implemented in @lucidrains 's x-transformers library on some toy experiments (pretraining a small model on [fineweb-edu](https://huggingface.co/datasets/HuggingFaceFW/fineweb-edu)); after I made sure that the baseline hparams match with the nGPT hparams (eg. enabling tied embeddings). I also did a small LR sweep, but I didn't train for very long, so my negative results may not transfer to a more scaled up setup.

**Comment by lucidrains** [comment]
> Hi Phil, I can now share what I did. It was for a Kaggle competition. I wrote a short paper, you can see the content there: https://www.kaggle.com/competitions/arc-prize-2024/discussion/545844 . I'll share my code later today.

very ambitious! did not expect you to be training on the ARC challenge

@JeanKaddour thanks for sharing your result as well, and i've actually read your optimizer paper! did you notice any signs of it being worse, or was your result about the same with non-normalized baseline?

**Comment by jfpuget** [comment]
I open sourced my solution at https://github.com/jfpuget/ARC-AGI-Challenge-2024

**Comment by JeanKaddour** [comment]
@lucidrains, I ran some more experiments, and they yielded some positive results for nGPT; in contrast to my previous ones.
The nGPT runs are in red, the baseline in green. 

<img width="1807" alt="image" src="https://github.com/user-attachments/assets/e6b7616d-94c0-4169-a7ca-b50270e0894e">
<img width="1805" alt="image" src="https://github.com/user-attachments/assets/2c96d53a-90af-49f9-8aee-1f88d39deb47">

* nGPT is more sensitive to the initial LR, ie there is higher variance across performance w.r.t. the LR. 
* For lr=0.001, it outperformed the baseline and overall resulted in the best performance across all runs within this sweep. 
* It also seems that for the baseline, there is a clear trend towards larger LRs resulting in better performance, and if we were to add a larger LR to the sweep, it might reach nGPT's performance. 

Happy to keep you updated with more results in the future.

**Comment by lucidrains** [comment]
@JeanKaddour yes, please do!

**Comment by lucidrains** [comment]
@JeanKaddour ah, it would seem as though they may have made a mistake in their baseline. someone has pointed out an error in their official repo

oh well, still an interesting circuit nonetheless!
